src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,387 @@
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Writes a cache entry to the store for the given Request and Response.
+     *
+     * Existing entries are read and any that match the response are removed. This
+     * method calls write with the new list of cache entries.
+     *
+     * @param Request  $request  A Request instance
+     * @param Response $response A Response instance
+     *
+     * @return string The key under which the response is stored
+     */
+    public function write(Request $request, Response $response)
+    {
+        if (!$response->headers->has('X-Content-Digest')) {
+            $contentDigest = $this->generateContentDigest($response);
+
+            if (false === $this->save($contentDigest, $response->getContent())) {
+                throw new \RuntimeException('Unable to store the entity.');
+            }
+
+            $response->headers->set('X-Content-Digest', $contentDigest);
+
+            if (!$response->headers->has('Transfer-Encoding')) {
+                $response->headers->set('Content-Length', strlen($response->getContent()));
+            }
+        }
+
+        $cacheKey = $this->getCacheKey($request);
+        $headers = $response->headers->all();
+        unset($headers['age']);
+
+        $item = $this->cache->getItem($cacheKey);
+
+        if (!$item->isHit()) {
+            $entries = [];
+        } else {
+            $entries = $item->get();
+        }
+
+        // Add or replace entry with current Vary header key
+        $entries[$this->getVaryKey($response->getVary(), $response->headers)] = [
+            'vary' => $response->getVary(),
+            'headers' => $headers,
+            'status' => $response->getStatusCode(),
+        ];
+
+        // If the response has a Vary header we remove the non-varying entry
+        if ($response->hasVary()) {
+            unset($entries[self::NON_VARYING_KEY]);
+        }
+
+        // Support tagging
+        $tags = [];
+        if ($response->headers->has($this->purgeTagsHeader)) {
+            $tags = explode(',', $response->headers->get($this->purgeTagsHeader));
+        }
+
+        $this->save($cacheKey, $entries, $tags);
+
+        return $cacheKey;
+    }
+
+    /**
+     * Invalidates all cache entries that match the request.
+     *
+     * @param Request $request A Request instance
+     */
+    public function invalidate(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        $this->cache->deleteItem($cacheKey);
+    }
+
+    /**
+     * Locks the cache for a given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool|string true if the lock is acquired, the path to the current lock otherwise
+     */
+    public function lock(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        if (isset($this->locks[$cacheKey])) {
+            return false;
```

Comment:
why not trying to acquire this lock ?

Reply:
Basically, I checked the implementation of `Store` (see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/HttpCache/Store.php#L69) so that's why it's the way it is. How should that whole method look like in your opinion?

Reply:
So if you lock then release. You'll never be able to lock again... Is it the expected behavior?

Reply:
I guess not :) You mean either we drop the `isset()` there or if we keep it as is, we need to `unset($this->locks[$cacheKey]);` in `unlock()`? :)

Reply:
Yep, either unset in the release method. Or try to acquire it here. As you want

Reply:
Fixed and covered by unit test in 5f7a1ea55b7592c3f8d0e3b3996a655ffbbb6edc.